Breaking change in TimeSpan fixed

tomasherceg opened the pull request:
We've introduced a breaking change in DotVVM 3.0 which causes JavaScript errors when `TimeSpan` is present in the viewmodel.

`TimeSpan` was not officially supported (and it still isn't) - we don't support formatting of `TimeSpan` values, and other things.

However, this PR "fixes" the behavior so `TimeSpan` can be used in the viewmodel and is not broken when sent back to the server.

exyi commented on src/DotVVM.Framework/Resources/Scripts/metadata/primitiveTypes.ts:
```diff
@@ -178,3 +178,17 @@ function validateDateTime(value: any) {
     }
 }
+
+function validateTime(value: any) {
```
```suggestion
function validateTimeSpan(value: any) {
```

exyi replied:
I think that calling it TimeSpan even on client-side will be clearer

exyi commented on src/DotVVM.Framework/Resources/Scripts/serialization/date.ts:
```diff
@@ -1,14 +1,30 @@
 import { logWarning } from "../utils/logging";
 
+const timeSpanBaseDate = new Date("0001-01-01 00:00:00.0000");
```
I don't understand what is going on, but I'm afraid this may break in some cases:
![image](https://user-images.githubusercontent.com/7894687/115057549-b8c2e300-9ed3-11eb-9918-dd79ee1020d9.png)

exyi replied:
Aha, the year 0001 is parsed in Node in non-UTC timezone as year 2001 :facepalm: Do we have this covered with (unit) tests?

exyi commented on src/DotVVM.Framework/Resources/Scripts/serialization/date.ts:
```diff
@@ -61,9 +77,22 @@ export function serializeTime(date: string | Date | null, convertToUtc: boolean
+    const hours = (ticks / 1000 / 3600) | 0;
+    const minutes = (ticks / 1000 / 60) | 0;
+    const seconds = (ticks / 1000) | 0;
+    const milliseconds = (ticks % 1000) | 0;
```
I'm not sure how often will this fail - esp. with regards to dayling saving time. Since it does not allow me to use 0001-01-01 as the base date, I use 2021-01-01. Your logic says that at 3:00 27.3. is h = 3, while at 3:00 28. 3. h = 2:

```


> (new Date("2021-03-28 03:00:00.0000").getTime() - new Date("2021-01-01 00:00:00.0000")) / 1000 / 3600 % 24
2
> (new Date("2021-03-27 03:00:00.0000").getTime() - new Date("2021-01-01 00:00:00.0000")) / 1000 / 3600 % 24
3
```

exyi replied:
Actually I don't know what should be the result...

exyi reviewed (commented):
I hate date and time. I don't think this works the same as did before.

tomasherceg commented:
I've reworked this to not depend on Date - it was a stupid idea in general.
I've also added JS tests so valid formats of timespan that need to pass the validation, are now covered.
